### ISSUE TITLE
review thread safety of client/API/Evaluation context

### ISSUE THREAD

**Issue by toddbaert** [open-feature/python-sdk#96]

**Comment by toddbaert** [comment]
@matthewelwell @hlipsig @tcarrio do we need to worry about locking/concurrency here? I know there's a `threading` module in Python. Are any locks necessary, especially around global state? From our meeting if sounds like this might not be necessary, but I have limited Python experience and can't say...

**Comment by beeme1mr** [comment]
Hey @federicobond perhaps you could answer this question. Do we need to be concerned about threading issues?

**Comment by federicobond** [comment]
I'd have to take a closer look at it. The GIL has us covered in some cases for simple assignments but anything more complex than that should probably be guarded by a lock.

`add_hooks` is probably not thread-safe due to read-update-write race conditions when concatenating lists.

`set_provider` used to be thread safe but likely isn't now with the new shutdown and initialize logic.

**Comment by toddbaert** [comment]
Did an audit of this as part of looking into 1.0 readiness (#522). Updated the description with findings.

tl;dr: the only locks in the SDK are in `_event_support.py`. Everything else (registry, hooks, eval context, transaction context, client state, provider `_on_emit`) is unprotected. The `ProviderRegistry` is the worst offender since it has multi-step mutations with side effects (init/shutdown), but there are races in most shared mutable state.

See the description above for the full list.

**Comment by leakonvalinka** [comment]
I'll start working on this now!
